black_it/loss_functions/likelihood.py

```diff
@@ -0,0 +1,133 @@
+# Black-box ABM Calibration Kit (Black-it)
+# Copyright (C) 2021-2022 Banca d'Italia
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""This module defines the 'LikelihoodLoss'  class."""
+import warnings
+from typing import Callable, List, Optional
+
+import numpy as np
+from numpy.typing import NDArray
+
+from black_it.loss_functions.base import BaseLoss
+
+
+def kernel(sq_dist: NDArray[np.float64], h: float, D: int) -> NDArray[np.float64]:
+    """Compute a kernel density estimation using a Gaussian kernel."""
+    k = np.exp(-(sq_dist / (2 * h**2))) / (h**D * (2 * np.pi) ** (D / 2.0))
+    return k
+
+
+class LikelihoodLoss(BaseLoss):
+    """Class for a likelihood loss computation using a Gaussian kernel function.
+
+    The likelihood computation can be useful within the approach proposed in
+    "Grazzini et al, Bayesian estimation of agent-based models, JEDC (2017)".
+
+    """
+
+    def __init__(
+        self,
+        coordinate_weights: Optional[NDArray] = None,
+        coordinate_filters: Optional[List[Optional[Callable]]] = None,
+        h: Optional[float] = None,
+    ):
+        """
+        Initialize the loss function.
+
+        Args:
+            coordinate_weights: the weights of the loss coordinates.
+            coordinate_filters: filters/transformations to be applied to each simulated series before
+                the loss computation.
+            h: the bandwidth of the kernel used to estimate the likelihood
+        """
+        super().__init__(coordinate_weights, coordinate_filters)
+        self.h = h
+
+    @staticmethod
+    def _get_bandwidth_silverman(N: int, D: int) -> float:
+        """Return a reasonable bandwidth value computed using the Silverman's rule of thumb."""
+        h = ((N * (D + 2)) / 4) ** (-1 / (D + 4))
+        return h
+
+    @staticmethod
+    def _get_bandwidth_scott(N: int, D: int) -> float:
```

Comment:
not used; maybe we can have a configuration in the loss (either scott or silverman)